opentelemetry-exporter-geneva/geneva-uploader/src/config_service/client.rs

```diff
@@ -0,0 +1,552 @@
+    ///   - `Identity`: Base64-encoded identity string (format: "Tenant=Default/Role=GcsClient/RoleInstance={agent_identity}")
+    ///   - `OSType`: Operating system type (Darwin/Windows/Linux)
+    ///   - `ConfigMajorVersion`: Version string (format: "Ver{major_version}v0")
+    ///   - `TagId`: UUID for request tracking
+    /// - **Headers**:
+    ///   - `User-Agent`: "{agent_identity}-{agent_version}"
+    ///   - `x-ms-client-request-id`: UUID for request tracking
+    ///   - `Accept`: "application/json"
+    ///
+    /// ## Response
+    /// - **Status**: 200 OK on success
+    /// - **Content-Type**: application/json
+    /// - **Body**:
+    ///   ```json
+    ///   {
+    ///     "IngestionGatewayInfo": {
+    ///       "endpoint": "https://ingestion.endpoint.example",
+    ///       "AuthToken": "auth-token-value"
+    ///     }
+    ///   }
+    ///   ```
+    ///
+    /// ## Authentication
+    /// Uses mutual TLS (mTLS) with client certificate authentication
+    ///
+    /// # Returns
+    /// * `Result<IngestionGatewayInfo, MonikerInfo>` - Ingestion gateway information, with storage monikers or an error
+    ///
+    /// # Errors
+    /// * `GenevaConfigClientError::Http` - If the HTTP request fails
+    /// * `GenevaConfigClientError::RequestFailed` - If the server returns a non-success status
+    /// * `GenevaConfigClientError::AuthInfoNotFound` - If the response doesn't contain ingestion info
+    /// * `GenevaConfigClientError::SerdeJson` - If JSON parsing fails
+    #[allow(dead_code)]
+    pub(crate) async fn get_ingestion_info(
+        &self,
+    ) -> Result<(IngestionGatewayInfo, MonikerInfo, String)> {
+        // First, try to read from cache (shared read access)
+        if let Ok(guard) = self.cached_data.read() {
+            if let Some(cached_data) = guard.as_ref() {
+                let expiry = cached_data.token_expiry;
+                if expiry > Utc::now() + chrono::Duration::minutes(5) {
+                    return Ok((
+                        cached_data.auth_info.0.clone(),
+                        cached_data.auth_info.1.clone(),
+                        cached_data.token_endpoint.clone(),
+                    ));
+                }
+            }
+        }
+
+        // Cache miss or expired token, fetch fresh data
+        // Perform actual fetch before acquiring write lock to minimize lock contention
+        let (fresh_ingestion_gateway_info, fresh_moniker_info) =
+            self.fetch_ingestion_info().await?;
+
+        let token_expiry = self
+            .parse_token_expiry(&fresh_ingestion_gateway_info.auth_token_expiry_time)
+            .ok_or_else(|| {
+                GenevaConfigClientError::InternalError("Failed to parse token expiry".into())
+            })?;
+
+        let token_endpoint = extract_endpoint_from_token(&fresh_ingestion_gateway_info.auth_token)
+            .map_err(|e| {
+                GenevaConfigClientError::Certificate(format!(
```

Comment:
This is not valid for `CertificateError`. In fact, there is no need to map the error. `extract_endpoint_from_token` method would only return `JwtTokenError` which is already a `GenevaCVonfigClientError` variant.